File: packages/gbnf-experiment/data/20260909T023701Z_e94dca40/ported_implementation/src/grammar-graph/get-parent-stack-id.ts
import { Color } from './colorize.js';
import type { GraphPointer } from './graph-pointer.js';

export const getParentStackId = (
  pointer: GraphPointer,
  col: (value: string | number, color: string) => string,
): string => {
  const stackIds: string[] = [];
  let parent: GraphPointer | undefined = pointer.parent;
  while (parent) {
    stackIds.push(
      `${parent.node.meta.stackId},${parent.node.meta.pathId},${parent.node.meta.stepId}`,
    );
    parent = parent.parent;
  }
  const arrow = col('<-', Color.GRAY);
  return stackIds.map((stackId) => col(stackId, Color.RED)).join(arrow);
};

File: proposed-projects/round-trip-experiment/reverse/20260909T002206Z_273cf6b9/20260910T042804Z_e72f6c04/ported_implementation/gbnf/grammar_graph/get_parent_stack_id.py
from typing import List

from .colorize import Color


def get_parent_stack_id(pointer, col) -> str:
    stack_ids: List[str] = []
    parent = pointer.parent
    while parent:
        stack_ids.append(
            f"{parent.node.meta.stack_id},{parent.node.meta.path_id},"
            f"{parent.node.meta.step_id}"
        )
        parent = parent.parent
    arrow = col("<-", Color.GRAY)
    return arrow.join(col(stack_id, Color.RED) for stack_id in stack_ids)
